### PR TITLE
#240 forgot/reset password

### LINKED ISSUE #240: Implement Forgot/Reset Password
Related Epic #185

### PR DESCRIPTION
**Pull request by colinvandergraaf**
- **change navigation route to on google auth sign in**
- **add send password link functionality; add snackbar toast messages**

## Description
Fixes: #240 
This PR creates the password reset functionality for users created with email and passwords.  

Fixes # (issue)

## Type of change

- [x] ✨  New feature (non-breaking change which adds functionality)
- [ ] Other 🗒️ ______________

## How Can this be tested? Testing Plan to review this PR

1.  Fire up the app and go to the `login` route and click on the `update` link next to forgot password. This takes you to the ‘forgotten’ route 
2. Enter a random email that doesn’t exist in our database.  See Snackbar notification for failing. 
3. Enter an email address that you have used to create a user in our App.  If you haven’t created one yet, go to create user and do so. 
4. Enter your email and see Snackbar messages for success and errors. 
5. Check your email for link and click on it.  This will take you back to the Linkta app to our `update-password` route.  
6. Type in a new password.  password must contain the follwoing: 1 uppercase letter, 1 lowercase letter, 1 number and 1 symbol.  Confirm your password in the Confirm password and submit.  
7. See SnackbarNotification for Success and be rerouted to `login` after 2.5 seconds

Currently there is no error handling for submitting an email that doesn’t exist in our database.  I tried using the `fetchSignInMethodsForEmail` - a firebase method intended for this, but this has been deprecated for for security  improvements around email enumeration. New projects created after 9/15/23 don’t have access to this function

[This article](https://medium.com/coffee-software/firebase-email-enumeration-enforcement-a-workaround-for-fetchsigninmethodsforemail-13a72a88f26e) suggests implementing this with a Firebase function.  I think this merits some looking into in a follow on ticket.  

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I have reviewed the [**Security Best Practices Document**](./docs/SECURITY_BEST_PRACTICES.md).
- [x] I have followed the [**Naming Conventions Guide**](./docs/NAMING_CONVENTIONS_GUIDE.md) for my changes.
- [x] I have ensured that my pull request title is descriptive.

### CODE REVIEW

**Review by relandboyle — APPROVED** [review]
@colinvandergraaf I went through the process and it all seems to work as it should. I tried to break it by using Google OAuth to login with an email that was already set up as a username/password, but it didn't break.

image

File: client/features/auth-pages/ForgotPasswordPage.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  } = useForm<FormData>({
+    resolver: zodResolver(userEmailSchema),
+  });
+
+  const { mutate, isPending, isSuccess } = useGetPasswordResetLinkMutation();
+
+  const [snackbarOpen, setSnackbarOpen] = useState(false);
+  const [snackbarMessage, setSnackbarMessage] = useState('');
+  const [snackbarSeverity, setSnackbarSeverity] = useState<
+    'error' | 'warning' | 'info' | 'success'
```
Comment (kinskiwu) [inline client/features/auth-pages/ForgotPasswordPage.tsx:33]: There is a SnackbarSeverity type in snackBar.d.ts that you may consider using:
```
import type { SnackbarSeverity } from '@/types/snackBar';

const [snackbarSeverity, setSnackbarSeverity] =
    useState<SnackbarSeverity>('info');
    
```

File: client/features/auth-pages/ForgotPasswordPage.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  >('info');
+
+  const handleCloseSnackbar = () => {
```
Comment (kinskiwu) [inline client/features/auth-pages/ForgotPasswordPage.tsx:36]: Maybe a good practice to also clear the snackbarMessage and snackbarSeverity state when closing the snack bar?

**Review by kinskiwu — APPROVED** [review]
Thanks for the good work! The password feature works as expected and I was able to submit the form with my keyboard, and I like how you enhanced the password strength too.

This is not related to the functionality of this feature, but for overall UX: as a user, I’d also like to have the option to see my new password when I reset it.

The current implementation is to redirect a user back to `login` after successful password update. From a UX perspective, I’m more inclined towards redirecting the user to `/generate` instead of the sign-in page since we have confirmed they are a verified user (and they came from the sign-in page originally).

One thing I noticed is that I was able to reset the password for my Google OAuth email. Do you happen to know why this could happen? Since I will never need to use a password for OAuth sign-in so it seems a little confusing that I get to reset my password.

**Comment by relandboyle** [comment]
> One thing I noticed is that I was able to reset the password for my Google OAuth email. Do you happen to know why this could happen? Since I will never need to use a password for OAuth sign-in so it seems a little confusing that I get to reset my password.

@kinskiwu If you've created an account using OAuth and 'email-address-1' and then you run through the password reset process using 'email-address-1' , the user record in Firebase Auth will be updated to be both an OAuth and an email/password account.

image

**Pull request merged**